crates/re_query_cache/src/flat_vec_deque.rs

```diff
@@ -0,0 +1,789 @@
+    #[inline]
+    pub fn range(&self, entry_range: Range<usize>) -> impl Iterator<Item = &[T]> {
+        let (values_left, values_right) = self.values.as_slices();
+        // NOTE: We can't slice into our offsets, we don't even know if they're contiguous in
+        // memory at this point -> skip() and take().
+        self.iter_offset_ranges()
+            .skip(entry_range.start)
+            .take(entry_range.len())
+            .map(|offsets| {
+                // NOTE: We do not need `make_contiguous` here because we always guarantee
+                // that a single entry's worth of values is fully contained in either the left or
+                // right buffer, but never straddling across both.
+                if offsets.start < values_left.len() {
+                    &values_left[offsets]
+                } else {
+                    &values_right[offsets]
+                }
+            })
+    }
+}
+
+#[test]
+fn range() {
+    let mut v: FlatVecDeque<i64> = FlatVecDeque::new();
+
+    assert_eq!(0, v.num_entries());
+    assert_eq!(0, v.num_values());
+
+    v.insert_range(0, [vec![1, 2, 3], vec![4, 5, 6, 7], vec![8, 9, 10]]);
+    assert_deque_eq(&[&[1, 2, 3], &[4, 5, 6, 7], &[8, 9, 10]], &v);
+
+    assert_iter_eq(&[&[1, 2, 3]], v.range(0..1));
+    assert_iter_eq(&[&[4, 5, 6, 7]], v.range(1..2));
+    assert_iter_eq(&[&[8, 9, 10]], v.range(2..3));
+
+    assert_iter_eq(
+        &[&[1, 2, 3], &[4, 5, 6, 7], &[8, 9, 10]],
+        v.range(0..v.num_entries()),
+    );
+
+    assert_iter_eq(&[&[1, 2, 3], &[4, 5, 6, 7], &[8, 9, 10]], v.iter());
+}
+
+// ---
+
+impl<T> FlatVecDeque<T> {
+    /// Prepends an entry comprised of `values` to the deque.
+    ///
+    /// This is the same as `self.insert(0, values)`.
+    ///
+    /// See [`Self::insert`] for more information.
+    #[inline]
+    pub fn push_front(&mut self, values: impl ExactSizeIterator<Item = T>) {
+        self.insert(0, values);
+    }
+
+    /// Appends an entry comprised of `values` to the deque.
+    ///
+    /// This is the same as `self.insert(self.num_entries(), values)`.
+    ///
+    /// See [`Self::insert`] for more information.
+    #[inline]
+    pub fn push_back(&mut self, values: impl ExactSizeIterator<Item = T>) {
+        self.insert(self.num_entries(), values);
+    }
+
+    /// Inserts a single entry at `entry_index`, comprised of the multiple elements given as `values`.
+    ///
+    /// This is O(1) if `entry_index` corresponds to either the start or the end of the deque.
+    /// Otherwise, this requires splitting the deque into two pieces then stitching them back together
+    /// at both ends of the added data.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    /// Panics if `values` is empty.
+    #[inline]
+    pub fn insert(&mut self, entry_index: usize, values: impl ExactSizeIterator<Item = T>) {
+        let num_values = values.len();
+        let deque = Self {
+            values: values.collect(),
```

Comment:
you can replace `ExactSizeIterator` with a normal iterator if you collect first, and then ask the length of the collected values.  The performance should be just as good thanks to `collect`s use of [`size_hint`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.size_hint)